onnx/defs/math/defs.cc

```diff
@@ -1665,4 +1665,54 @@ ONNX_OPERATOR_SET_SCHEMA(
           }
         }));
 
+// Is this operator should be a function? (see https://github.com/onnx/onnx/pull/2242)
+// The operator can be replaced by a function but this function
+// is different for every supported metric.
+static const char* CDist_ver12_doc = R"DOC(
+Compute pair-wise distances between two 2D tensors
+sharing the same number of columns. The output
+is a 2D symmetric float matrix.
+
+Example:
+```
+input_x = [[1, 2, 3], [4, 5, 6]]
+input_y = [[7, 8, 9]]
+metric = 'sqeuclidean'
+output = [[108], [27]]
+```
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    CDist,
+    12,
+    OpSchema()
+        .SetDoc(CDist_ver12_doc)
+        .Input(0, "X", "Input tensor X. Shape of X should be (M, K).", "T")
+        .Input(1, "Y", "Input tensor B. Shape of Y should be (N, K).", "T")
+        .Output(0, "Z", "Output tensor of shape (M, N).", "T")
+        .TypeConstraint(
+            "T", {"tensor(float16)", "tensor(float)", "tensor(double)"},
+            "Constrain input and output types to float tensors.")
+        .Attr("metric",
+              "Metric to use to compute distances, it can be euclidean, "
```

Comment:
Scipy supports much more metrics. Do we need any of these in the future? If yes, we'd better to add them here now.

Reply:
I kept the list to what is commonly used. Here is the whole list:
```
'seuclidean' 'cosine' 'correlation' 'hamming' 'jaccard'
'chebyshev' 'canberra' 'braycurtis' 
'yule' 'dice' 'kulsinski' 'rogerstanimoto' 'russellrao' 'sokalmichener' 'sokalsneath'
'mahalanobis' 'wminkowski' 'haversine'
```
See [cdist](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html), and [DistanceMetric](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.DistanceMetric.html?highlight=distancemetric#sklearn.neighbors.DistanceMetric).

So you think I should do all of them?

Reply:
@linkerzhang @ebarsoum could you help review this op and give some guidance here?

Reply:
I suggest we put all of them here, to avoid bumping up the op version in the future.

Reply:
Before I add them, what do you think of the rest of the PR? Is this the only required change?